assembly/BigInt.ts

```diff
@@ -815,7 +815,167 @@ export class BigInt {
 
   // EXPONENTIATION ////////////////////////////////////////////////////////////////////////////////////////////////////
 
-  pow(k: i32): BigInt {
+  pow<T>(val: T): BigInt {
+    if (val instanceof BigInt) return this.powBigint(val);
+    // @ts-ignore
+    if (val instanceof i8) return this.powInt8(val);
+    // @ts-ignore
+    if (val instanceof u8) return this.powUint8(val);
+    // @ts-ignore
+    if (val instanceof i16) return this.powInt16(val);
+    // @ts-ignore
+    if (val instanceof u16) return this.powUint16(val);
+    // @ts-ignore
+    if (val instanceof i32) return this.powInt32(val);
+    // @ts-ignore
+    if (val instanceof u32) return this.powUint32(val);
+    // @ts-ignore
+    if (val instanceof i64) return this.powInt64(val);
+    // @ts-ignore
+    if (val instanceof u64) return this.powUint64(val);
+    throw new TypeError("Unsupported generic type " + nameof<T>(val));
+  }
+
+  powBigint(k: BigInt): BigInt {
+    if (k.isNeg) {
+      throw new RangeError("BigInt does not support negative exponentiation");
+    }
+    if (k.lte(BigInt.fromUInt64(<u64>F64.MAX_SAFE_INTEGER))) {
+      return this.powUint64(k.toUInt64());
+    }
+    let temp: BigInt = this.copy();
+    let res: BigInt = BigInt.ONE;
+    while (k.lt(BigInt.ZERO)) {
+      /* if the bit is set multiply */
+      if (k.bitwiseAnd(1).ne(0)) res = res.mul(temp);
+      /* square */
+      if (k.gt(BigInt.ONE)) temp = temp.square();
+      /* shift to next bit */
+      k = k.rightShift(1);
+    }
+    return res;
+  }
+
+  powUint32(k: u32): BigInt {
```

Comment:
I don't like how we're duplicating the code manually here for all integer cases, the AssemblyScript templating engine should handle this for us ideally. Are AssemblyScript templates really this limited in their capabilities?

Reply:
Amazing, thank you!